Mathlib/Dynamics/TopologicalEntropy/CoverEntropy.lean

```diff
@@ -0,0 +1,617 @@
+
+lemma IsDynCoverOf.nonempty_inter {T : X → X} {F : Set X} {U : Set (X × X)} {n : ℕ} {s : Finset X}
+    (h : IsDynCoverOf T F U n s) :
+    ∃ t : Finset X, IsDynCoverOf T F U n t ∧ t.card ≤ s.card
+    ∧ ∀ x ∈ t, ((ball x (dynEntourage T U n)) ∩ F).Nonempty := by
+  classical
+  use Finset.filter (fun x : X ↦ ((ball x (dynEntourage T U n)) ∩ F).Nonempty) s
+  simp only [Finset.coe_filter, Finset.mem_filter, and_imp, imp_self, implies_true, and_true]
+  apply And.intro _ (Finset.card_mono (Finset.filter_subset _ s))
+  intros y y_F
+  specialize h y_F
+  simp only [Finset.coe_sort_coe, mem_iUnion, Subtype.exists, exists_prop] at h
+  rcases h with ⟨z, z_s, y_Bz⟩
+  simp only [coe_setOf, mem_setOf_eq, mem_iUnion, Subtype.exists, exists_prop]
+  use z
+  exact ⟨⟨z_s, nonempty_of_mem ⟨y_Bz, y_F⟩⟩, y_Bz⟩
+
+/-This lemma is the first step in a submultiplicative-like property of `coverMincard`, with
+  far-reaching consequences such as explicit bounds for the topological entropy
+  (`coverEntropyInfUni_le_card_div`) and an equality between two notions of topological entropy
+  (`coverEntropyInf_eq_coverEntropySup_of_inv`).-/
+lemma IsDynCoverOf.iterate_le_pow {T : X → X} {F : Set X} (F_inv : MapsTo T F F) {U : Set (X × X)}
+    (U_symm : SymmetricRel U) {m : ℕ} (n : ℕ) {s : Finset X} (h : IsDynCoverOf T F U m s) :
+    ∃ t : Finset X, IsDynCoverOf T F (U ○ U) (m * n) t ∧ t.card ≤ s.card ^ n := by
+  classical
+  rcases F.eq_empty_or_nonempty with (rfl | F_nemp)
+  · use ∅; simp
+  have _ : Nonempty X := nonempty_of_exists F_nemp
+  have s_nemp := h.nonempty F_nemp
+  rcases F_nemp with ⟨x, x_F⟩
+  rcases m.eq_zero_or_pos with (rfl | m_pos)
+  · use {x}
+    simp only [zero_mul, Finset.coe_singleton, Finset.card_singleton]
+    exact And.intro (isDynCoverOf_zero T F (U ○ U) (singleton_nonempty x))
+      <| one_le_pow_of_one_le' (Nat.one_le_of_lt (Finset.Nonempty.card_pos s_nemp)) n
+  have : ∀ β : Fin n → s, ∃ y : X, (⋂ k : Fin n, T^[m * k] ⁻¹' ball (β k) (dynEntourage T U m)) ⊆
+      ball y (dynEntourage T (U ○ U) (m * n)) := by
+    intro t
+    rcases (⋂ k : Fin n, T^[m * k] ⁻¹' ball (t k) (dynEntourage T U m)).eq_empty_or_nonempty
+      with (inter_empt | inter_nemp)
+    · rw [inter_empt]
+      use x
+      exact empty_subset _
+    · rcases inter_nemp with ⟨y, y_int⟩
+      use y
+      intro z z_int
+      simp only [ball, dynEntourage, Prod.map_iterate, mem_preimage, mem_iInter,
+        Prod.map_apply] at y_int z_int ⊢
+      intro k k_mn
+      replace k_mn := Nat.div_lt_of_lt_mul k_mn
+      specialize z_int ⟨(k / m), k_mn⟩ (k % m) (Nat.mod_lt k m_pos)
+      specialize y_int ⟨(k / m), k_mn⟩ (k % m) (Nat.mod_lt k m_pos)
+      rw [← Function.iterate_add_apply T (k % m) (m * (k / m)), Nat.mod_add_div k m] at y_int z_int
+      exact mem_comp_of_mem_ball U_symm y_int z_int
+  choose! dyncover h_dyncover using this
+  let sn := range dyncover
+  have _ := fintypeRange dyncover
+  use sn.toFinset
+  constructor
```

Comment:
```suggestion
  refine ⟨sn.toFinset, ?_, ?_⟩
```

Reply:
You resolved the conversation, but I don't see the change. Did you forget to push?

Reply:
Sorry for this, pushed now.